File: backend/app/routers/clients.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from .. import crud, schemas
from ..database import get_db

router = APIRouter(prefix="/api/clients", tags=["clients"])


@router.get("", response_model=List[schemas.Client])
def list_clients(db: Session = Depends(get_db)):
    return crud.get_clients(db)


@router.post("", response_model=schemas.Client, status_code=201)
def create_client(body: schemas.ClientCreate, db: Session = Depends(get_db)):
    return crud.create_client(db, body)


@router.get("/{id}", response_model=schemas.Client)
def get_client(id: int, db: Session = Depends(get_db)):
    obj = crud.get_client(db, id)
    if not obj:
        raise HTTPException(status_code=404, detail="Client not found")
    return obj


@router.put("/{id}", response_model=schemas.Client)
def update_client(id: int, body: schemas.ClientCreate, db: Session = Depends(get_db)):
    obj = crud.update_client(db, id, body)
    if not obj:
        raise HTTPException(status_code=404, detail="Client not found")
    return obj


@router.delete("/{id}", status_code=204)
def delete_client(id: int, db: Session = Depends(get_db)):
    if not crud.delete_client(db, id):
        raise HTTPException(status_code=404, detail="Client not found")

File: frontend/src/api/clients.ts
import { apiClient as api } from '@/api/client'
import type { Client, ClientCreate } from '@/types/client'


export const getClients = (): Promise<Client[]> =>
  api.get<Client[]>('/api/clients').then(r => r.data)

export const getClient = (id: number): Promise<Client> =>
  api.get<Client>(`/api/clients/${id}`).then(r => r.data)

export const createClient = (body: ClientCreate): Promise<Client> =>
  api.post<Client>('/api/clients', body).then(r => r.data)

export const updateClient = (id: number, body: ClientCreate): Promise<Client> =>
  api.put<Client>(`/api/clients/${id}`, body).then(r => r.data)

export const deleteClient = (id: number): Promise<void> =>
  api.delete(`/api/clients/${id}`).then(() => undefined)
